modules/currency-l1/src/main/scala/io/constellationnetwork/currency/l1/http/DataApplicationRoutes.scala

```diff
@@ -47,55 +50,126 @@ final case class DataApplicationRoutes[F[_]: Async: Hasher: SecurityProvider: L1
+      implicit val dataUpdateEncoder: Encoder[DataUpdate] = dataApplication.dataEncoder
+      dataTransactionsQueue.view.flatMap(Ok(_))
 
     case req @ POST -> Root / "data" =>
-      implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
-      req
-        .asR[Signed[DataUpdate]] {
-          _.toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _).flatMap {
-            case Left(_) => BadRequest(InvalidSignature.toApplicationError)
-            case Right(hashed) =>
-              validate(hashed.signed) {
-                dataUpdatesQueue.offer(hashed.signed) >> Ok(("hash" ->> hashed.hash.value) :: HNil)
+      def parseMultipleDataTransactions: F[DataTransactions] = {
+        implicit val dataTransactionRequest: EntityDecoder[F, DataTransactions] = dataApplication.requestDecoder
+        req.as[DataTransactions]
+      }
+
+      def parseSingleDataTransaction: F[Signed[DataUpdate]] = {
+        implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
+        req
+          .as[Signed[DataUpdate]]
+      }
+
+      def handleDataUpdates(dataTransactions: DataTransactions): F[Response[F]] =
+        validateDataTransactions(dataTransactions) {
+          dataTransactions.traverse { signedTransaction =>
+            signedTransaction.value match {
+              case dataUpdate: DataUpdate =>
+                Signed(dataUpdate, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _)
+                  .map(_.leftMap(_.widen[DataTransaction]))
+                  .map(_.widen[Hashed[DataTransaction]])
+
+              case feeTransaction: FeeTransaction =>
+                Signed(feeTransaction, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](serialize[F] _)
+                  .map(_.leftMap(_.widen[DataTransaction]))
+                  .map(_.widen[Hashed[DataTransaction]])
+            }
+          }.flatMap { results =>
+            val (errors, validTransactions) = results.toList.partitionMap(identity)
+            if (errors.nonEmpty) {
+              BadRequest(
+                Json.obj("error" -> Json.fromString("Invalid signature in data transactions"))
+              )
+            } else {
+              NonEmptyList.fromList(validTransactions) match {
+                case None =>
+                  BadRequest(Json.obj("error" -> Json.fromString("No valid data transactions found")))
+                case Some(txs) =>
+                  dataTransactionsQueue
+                    .offer(txs.map(_.signed))
+                    .flatMap(_ => mkResponse(txs))
               }
+            }
           }
         }
+
+      def mkResponse(transactions: NonEmptyList[Hashed[DataTransaction]]): F[Response[F]] = {
+        val (dataUpdateHashes, feeTransactionHashes) = transactions.toList.partition(_.signed.value.isInstanceOf[DataUpdate])
```

Comment:
FYI - please remember that it works for two types of DataTransaction, but if we would create a third type, it would break this part silently

Reply:
I'm not sure, just thing to remember. but maybe the exhaustivity checking would help, so instead of partitioning like `is it an data update or anything else` -> `is it a data update or is it a fee transaction?` when adding new type, compiler would throw that there is one case not covered